design/one-pager-strongly-typed-class.md

```diff
@@ -0,0 +1,264 @@
+
+## Proposal
+
+Unfortunately, strongly typed resource classes make it much more difficult to identify default resource classes. Currently, there is one default class controller per claim kind that watches for creation of their specified claim kind, then lists objects of `kind: ResourceClass` that are labeled as default for the claim kind. With an unknown number of possible kinds of resource classes that could be installed and serve as default for a claim `kind`, a default class controller can no longer list a single `ResourceClass` kind.
+
+One solution would instead be to map resource classes one-to-one with claim kinds (i.e. `MySQLInstanceClass`), but that results in the same issue of varying required parameters that is currently being experienced.
+
+In order to satisfy both the requirements of strong typing and default resource classes, a new set of CRD's can be implemented to specify when an instance of a strongly typed resource class kind should serve as default for a given claim kind. These default policies should be able to be defined at both the cluster and namespace level, so there must exist two separate CRD's for each claim kind. For example, the `MySQLInstance` claim kind will be accompanied by the `MySQLInstancePolicy` (namespace-scoped) and `MySQLInstanceClusterPolicy` (cluster-scoped) kinds. Importantly, when single instances of both the `MySQLInstancePolicy` and `MySQLInstanceClusterPolicy` exist, the `MySQLInstancePolicy` would take precedence because it is at the more granular (namespace) level.
+
+## Workflow
+
+The workflow for a namespace-scoped policy (`kind: MySQLInstancePolicy`) would proceed as follows:
+
+1. An admin installs Crossplane, which specifies the portable claim kinds that are available as part of the core Crossplane.
+2. An admin installs an infrastructure stack (i.e. a set of CRD's and controllers for a provider's managed service offerings).
+3. An admin creates an instance of an installed resource class CRD (i.e. `kind: RDSInstanceClass`). It could look as follows:
+
+```yaml
+apiVersion: database.aws.crossplane.io/v1alpha1
+kind: RDSInstanceClass
+metadata:
+  name: standard-mysql
+  namespace: crossplane-system
+specTemplate:
+  class: db.t2.small
+  masterUsername: masteruser
+  securityGroups:
+    - sg-ab1cdefg
+    - sg-05adsfkaj1ksdjak
+  size: 20
+  providerRef:
+    name: aws-provider
+  reclaimPolicy: Delete
+```
+
+4. An admin creates a new `MySQLInstancePolicy` object, specifying that the previously created resource class (`kind: RDSInstanceClass`) instance should serve as default for a given claim kind `mysqlinstance.storage.crossplane.io`. A `MySQLInstancePolicy` could look as follows:
+
+```yaml
+apiVersion: core.crossplane.io/v1alpha1
```

Comment:
I think this would fit better in the same API group and version as `MySQLInstance`, i.e. `database.crossplane.io/v1alpha1`.

Reply:
Agreed!